src/client/opamClient.ml

```diff
@@ -1130,38 +1130,53 @@ let install t ?autoupdate ?add_to_roots
 
 let check_installed t atoms =
   let available = (Lazy.force t.available_packages) in
-  let pkgs = OpamFormula.packages_of_atoms available atoms in
+  let uninstalled = OpamPackage.Set.Op.(available -- t.installed) in
+  let pkgs =
+    OpamPackage.to_map
+      (OpamFormula.packages_of_atoms available atoms)
+  in
   let test = OpamStateConfig.(!r.build_test) in
   let doc = OpamStateConfig.(!r.build_doc) in
   let env p =
     OpamFilter.deps_var_env ~build:true ~post:true ~test ~doc
       ~dev:(OpamSwitchState.is_dev_package t p)
   in
-  OpamPackage.Set.fold (fun to_check_pkg map ->
-      let cnf_formula =
-        OpamSwitchState.opam t to_check_pkg
-        |> OpamFile.OPAM.depends
-        |> OpamFilter.filter_formula (env to_check_pkg)
-        |> OpamFormula.to_cnf
-      in
-      let missing_conj =
-        List.filter
-          (List.for_all (fun ((n,_vc) as atom) ->
-               OpamPackage.Set.for_all
-                 (fun p -> not (OpamFormula.check atom p))
-                 (OpamPackage.packages_of_name t.installed n)))
-          cnf_formula
+  OpamPackage.Name.Map.fold (fun name versions map ->
+      let compliant, missing_opt =
+        OpamPackage.Version.Set.fold (fun version (found, missing) ->
```

Comment:
If I am not mistaken, `found = true` ⇔ `missing = None`, so maybe this can be slightly simplified ?

Reply:
It would be cleaner on the idea, but more _complex_ code without the `found` argument: `None` can't be the first accumulator value (ends on the first call), in that case, as `missing` contains a package, first value need to be calculated by taking an element of the set & apply part of the fold arg function on it.

Reply:
Aah, good point indeed. You could start with the acc `Some []`, though.

Reply:
not really, because `missing: (package * atom disjunction list) option` (and there is no empty package) :)

Reply:
Right :)